src/vidpf.rs

```diff
@@ -0,0 +1,743 @@
+pub(crate) enum ServerId {
+    /// S0 is the first server.
+    S0,
+    /// S1 is the second server.
+    S1,
+}
+
+impl From<ServerId> for Choice {
+    fn from(value: ServerId) -> Self {
+        match value {
+            ServerId::S0 => Self::from(0),
+            ServerId::S1 => Self::from(1),
+        }
+    }
+}
+
+/// [`CorrectionWord`] adjusts values of shares during the VIDPF evaluation.
+struct CorrectionWord<W: VidpfWeight> {
+    seed: Seed,
+    left_control_bit: Choice,
+    right_control_bit: Choice,
+    weight: W,
+}
+
+/// [`PublicShare`] is the common public information used by aggregation servers.
+pub struct PublicShare<W: VidpfWeight> {
+    cw: Vec<CorrectionWord<W>>,
+    cs: Vec<Proof>,
+}
+
+/// [`EvalState`] contains the values produced during input evaluation at a given level.
+pub struct EvalState {
+    seed: Seed,
+    control_bit: Choice,
+    proof: Proof,
+}
+
+impl EvalState {
+    fn init_from_key(key: &Key) -> Self {
+        Self {
+            seed: key.value,
+            control_bit: Choice::from(key.id),
+            proof: Proof::default(),
+        }
+    }
+}
+
+/// [`ValueShare`] contains a share of the input's weight together with a proof for verification.
+pub struct ValueShare<W: VidpfWeight> {
+    /// Secret share of the input's weight.
+    pub share: W,
+    /// Proof used to verify the share.
+    pub proof: Proof,
+}
+
+/// Proof size in bytes.
+const VIDPF_PROOF_SIZE: usize = 32;
+
+/// [`Proof`] is used to validate user input and shares after evaluation.
+#[derive(Debug, PartialEq, Eq, Default)]
+pub struct Proof([u8; VIDPF_PROOF_SIZE]);
+
+impl Proof {
+    fn xor(mut self, other: &Self) -> Self {
+        zip(&mut self.0, other.0).for_each(|(a, b)| a.bitxor_assign(b));
+        self
+    }
+
+    fn conditional_xor(mut self, choice: Choice, other: &Self) -> Self {
+        zip(&mut self.0, other.0).for_each(|(a, b)| a.conditional_assign(&a.bitxor(b), choice));
+        self
+    }
+}
+
+/// [`Seed`] is used to feed a pseudorandom generator during evaluation.
+type Seed = [u8; 16];
+
+struct PrgOutput {
+    left_seed: Seed,
+    left_control_bit: Choice,
+    right_seed: Seed,
+    right_control_bit: Choice,
+}
+
+/// [`Weight`] represents an array of field elements that implements the [`VidpfWeight`] trait.
+#[derive(Debug, PartialEq, Eq)]
+pub struct Weight<F: FieldElement>(Vec<F>);
+
+impl<F: FieldElement> Weight<F> {
+    /// Creates a [`Weight`] from an array of field elements with length specified in parameters.
+    pub fn new_with_params(params: &WeightParameter<F>, array: &[F]) -> Result<Self, VidpfError> {
+        if params.length != array.len() {
+            return Err(VidpfError::InvalidWeightLength);
+        }
+
+        let slice = array
+            .get(0..params.length)
+            .ok_or(VidpfError::InvalidWeightLength)?;
+
+        Ok(Self(Vec::from(slice)))
+    }
+
+    /// Panics if weight lengths are different.
+    fn check_eq_length(a: &Self, b: &Self) {
```

Comment:
Code smell: It's a bit odd to have a function whose only function is to panic.

It looks like you're trying to enforce a precondition in certain places: that the inputs to a function have the same length.

First, it would be a bit more clear if you replaced `Self::chunk_eq_length(a b)` with `assert_eq!(a, b, "vidpf: weights with different number of field elements");`

Second, I wonder if we should use `debug_assert_eq!()` instead of `assert_eq!()` such that the condition is not evaluated in the release code.

Reply:
This condition needs to be evaluated on runtime; otherwise, without the checking, calculations will be performed incorrectly without noticing.
Open to other suggestions.

Reply:
I would just replace `Self::chunk_eq_length(a, b)` with `assert_eq!(a,b)`.